### PR TITLE
Use akka-http for requests from the invoker to the containers

### PR DESCRIPTION
**Pull request by psuter**
Stepping stone towards streaming `exec` data from the database to the containers. This commit replicates existing behavior but replaces the blocking Apache HTTP client by akka-http.

This commit also introduces a new helper object, `NewHttpUtils` to make non-pooled, possibly retrying, HTTP requests using the `akka-http` data model. (Candidates for a better object name are welcome.)

FYI @rabbah @markusthoemmes

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestClient.scala [hunk 1]
```diff
@@ -134,9 +134,8 @@ class CouchDbRestClient protected(system: ActorSystem, protocol: String, host: S
                     Unmarshal(response.entity.withoutSizeLimit()).to[JsObject].map { o => Right(o) }
                 } else {
                     // This is important, even though the response is ignored:
-                    // Unmarshalling is one way to drain the entity stream.
                     // Otherwise the connection stays open and the pool dries up.
-                    Unmarshal(response.entity).to[Array[Byte]].map { _ => Left(response.status) }
+                    response.entity.dataBytes.runWith(Sink.ignore).map { _ => Left(response.status) }
```
Comment (psuter) [inline common/scala/src/main/scala/whisk/core/database/CouchDbRestClient.scala]: This change is not directly related. @markusthoemmes couldn't help but notice I was wasting precious bytes here by keeping a copy of responses while ignoring them immediately after.

File: common/scala/src/main/scala/whisk/common/NewHttpUtils.scala [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import akka.actor.ActorSystem
+import akka.http.scaladsl._
+import akka.http.scaladsl.model._
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl._
+
+import java.util.concurrent.TimeoutException
+
+object NewHttpUtils {
+    def singleRequestBlocking(
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/common/NewHttpUtils.scala]: Do we even want to provide that one?

Comment (psuter) [inline common/scala/src/main/scala/whisk/common/NewHttpUtils.scala]: I'm using it in some subsequent work (getting rid of `ConsulKV`), but that will probably change too, so happy to remove.

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/common/NewHttpUtils.scala]: I see, I'ld prefer an explicit `Await` then though. This one hides the badness (for those who don't immediatly cringe at the word "blocking")

File: tests/src/actionContainers/ActionContainer.scala [hunk 3]
```diff
@@ -123,9 +123,9 @@ object ActionContainer {
     }
 
     private def syncPost(host: String, endPoint: String, content: JsValue): (Int, Option[JsObject]) = {
-        import whisk.common.HttpUtils
-        val connection = HttpUtils.makeHttpClient(30000, true)
-        val (code, bytes) = new HttpUtils(connection, host).dopost(endPoint, content, Map.empty)
+        import whisk.common.NewHttpUtils
```
Comment (markusthoemmes) [inline tests/src/actionContainers/ActionContainer.scala]: Does that even exist anymore?

**Comment by csantanapr** [comment]
@psuter What's the status on this PR? waiting for review, abandoned, ready?

**Comment by rabbah** [comment]
This is active. A commit was pushed 2 hours ago.

**Comment by csantanapr** [comment]
My bad, was sorting PRs my active, I guess I picked from the wrong end, and just saw the "21 days ago"

File: tests/src/actionContainers/ActionContainer.scala (hunk 3 above)
Comment (psuter) [inline tests/src/actionContainers/ActionContainer.scala]: @markusthoemmes  Panicked push to get the build starting before the tunnel :)

**Comment by psuter** [comment]
To answer @csantanapr's question, the reason I haven't asked for a merge yet is because, up to some days ago, occasionally a failure showed up in the build (not killing a long running action fast enough). I never witnessed the failure locally. I have been trying to reproduce it, but now it seems it just won't show up again.

**Comment by csantanapr** [comment]
Thanks for the update @psuter 👍🏼

**Comment by psuter** [comment]
PG1 90

**Comment by psuter** [comment]
PG1 91 as well.  Don't know how else to test at this point.

**Comment by csantanapr** [comment]
@perryibm @psuter ping, PR still under review?

**Comment by perryibm** [comment]
It's not a matter of code review but confidence due to the observation of failure and then inability to reproduce.  I suggest just running say 5 PGs to increase confidence and then we can merge and see what happens under load.

**Comment by psuter** [comment]
Thanks, will do.

**Comment by psuter** [comment]
@perryibm see comment above for a list of successful PG1 builds.

File: core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala [hunk 4]
```diff
@@ -127,27 +139,59 @@ class WhiskContainer(
...
+        import system.dispatcher
+
         val start = ContainerCounter.now()
-        val result = containerIP map { host =>
-            try {
-                val connection = HttpUtils.makeHttpClient(timeout, true)
-                val http = new HttpUtils(connection, host)
-                val (code, bytes) = http.dopost(endpoint, msg, Map(), timeout)
-                Try { connection.close() }
-                val returnCode = if (code == -1) ActivationResponse.ContainerError else code
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala]: This change (for the code == -1 on timeout) was not semantically preserving.

File: core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala [hunk 5]
```diff
@@ -127,27 +139,59 @@ class WhiskContainer(
...
+            for (
+                entity <- Marshal(msg).to[MessageEntity];
+                request = HttpRequest(method = HttpMethods.POST, uri = uri, entity = entity);
+                response <- NewHttpUtils.singleRequest(request, timeout, retryOnTCPErrors = true, retryInterval = 100.milliseconds);
+                responseBody <- Unmarshal(response.entity).to[String]
+            ) yield {
+                Some((response.status.intValue, responseBody))
+            }
+        } getOrElse {
+            Future.successful(None)
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala]: related to comment above.

File: core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala (hunk 4 above)
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/WhiskContainer.scala]: change might not be harmful but I think makes this statement in the invoker out of date:
`                    val removeContainer = failedInit || response.exists { _._1 == ActivationResponse.ContainerError }`

since the second term cannot be true anymore.
confirm? @markusthoemmes @psuter
